### PR TITLE
Scheduler

### PR DESCRIPTION
**Pull request by davidkpiano**

### CODE REVIEW

File: packages/core/src/actions/raise.ts [hunk 1]
```diff
@@ -85,7 +85,8 @@ function executeRaise(
   }
 ) {
   if (typeof params.delay === 'number') {
```
Comment (Andarist) [inline packages/core/src/actions/raise.ts:87]: I'd consider wrapping this in `defer` (ain't sure yet though)

File: packages/core/src/interpreter.ts [hunk 2]
```diff
@@ -150,7 +142,12 @@ export class Actor<TLogic extends AnyActorLogic>
     const { clock, logger, parent, syncSnapshot, id, systemId, inspect } =
       resolvedOptions;
 
-    this.system = parent?.system ?? createSystem(this);
+    this.system =
+      parent?.system ??
+      createSystem(this, {
+        clock,
+        snapshot: (options as any)?.snapshot?.['xstate.system']
```
Comment (Andarist) [inline packages/core/src/interpreter.ts:149]: for the root we probably should "split" the `options.snapshot` so `restoreSnapshot` of the root logic doesn't receive this `xstate.system` key there. We need to contain all of this attaching/detaching of the `xstate.system` magical thing in the Actor

File: packages/core/src/scheduler.ts [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+      to?: AnyActorRef;
+    }
+  ): void;
+  cancel(id: string): void;
+  cancelAll(actorRef: AnyActorRef): void;
+  getPersistedSnapshot(): Record<string, ScheduledEvent>;
+  start(): void;
+}
+
+const defaultClock: Clock = {
```
Comment (Andarist) [inline packages/core/src/scheduler.ts:35]: this is unused currently if I read it right

File: packages/core/src/scheduler.ts [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+      source: AnyActorRef,
+      data: {
+        id: string | undefined;
+        event: EventObject;
+        delay: number;
+        to: AnyActorRef;
+      }
+    ) => {
+      // TODO: `id` has to be separated from `data.id` completely
+      // `data.id` is supposed to be unique within an actor, `id` is supposed to be unique globally
```
[the remark this answers is not in the document]
Comment (Andarist) [reply packages/core/src/scheduler.ts:63]: we need tests for this :P

File: packages/core/src/scheduler.ts [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+        id,
+        source,
+        target: data.to || source,
+        startedAt: Date.now()
+      };
+      scheduler.events[id] = scheduledEvent;
+
+      const timeout = clock.setTimeout(() => {
+        const target = data.to || source;
+        // TODO: explain this hack, it should also happen sooner, not within this timeout
```
Comment (Andarist) [inline packages/core/src/scheduler.ts:76]: this likely is fixable (but I didn't check it) by deferring `scheudle`+`cancel` in our machine actions

File: packages/core/src/scheduler.ts [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+        }
+      }
+    },
+    getPersistedSnapshot: () => {
+      return { ...scheduler.events };
+    },
+    start: () => {
+      for (const id in scheduler.events) {
+        const scheduledEvent = scheduler.events[id];
+        scheduler.schedule(scheduledEvent.source, scheduledEvent);
```
Comment (Andarist) [inline packages/core/src/scheduler.ts:108]: TODO: we should diff the `now - createdAt +-/ delayThing`

Comment (Andarist) [reply packages/core/src/scheduler.ts:108]: an option for the behavior? but which one is the default?

File: packages/core/src/interpreter.ts [hunk 7]
```diff
@@ -695,7 +660,15 @@ export class Actor<TLogic extends AnyActorLogic>
...
   public getPersistedSnapshot(options?: unknown): Snapshot<unknown> {
-    return this.logic.getPersistedSnapshot(this._snapshot, options);
+    const persisted = this.logic.getPersistedSnapshot(this._snapshot, options);
+    if (!this._parent) {
+      const rootPersistedSnapshot = {
+        ...persisted,
+        'xstate.system': this.system.getPersistedSnapshot()
+      };
+      return rootPersistedSnapshot;
+    }
```
Comment (Andarist) [inline packages/core/src/interpreter.ts:670]: I'd temporarily disable this part of the code to figure out system-wide rehydration of actor references.

File: packages/core/src/interpreter.ts (hunk 2 above)
Comment (davidkpiano) [reply packages/core/src/interpreter.ts]: This part is removed

File: packages/core/src/interpreter.ts (hunk 7 above)
Comment (davidkpiano) [reply packages/core/src/interpreter.ts]: This part is removed

File: packages/core/src/system.ts [hunk 8]
```diff
@@ -1,22 +1,109 @@
...
+   */
+  _relay: (
+    source: AnyActorRef | undefined,
+    target: AnyActorRef,
+    event: AnyEventObject
+  ) => void;
+  scheduler: Scheduler;
+  getSnapshot: () => {
+    _scheduledEvents: Record<string, ScheduledEvent>;
+  };
```
Comment (Andarist) [inline packages/core/src/system.ts:75]: we'll likely need `getPersistedSnapshot` here too - but that can be added when we manage to implement serialization/deserialization of actor refs

File: packages/core/src/system.ts [hunk 9]
```diff
@@ -1,22 +1,109 @@
...
+
+export type AnyActorSystem = ActorSystem<any>;
+
 let idCounter = 0;
 export function createSystem<T extends ActorSystemInfo>(
-  rootActor: AnyActorRef
+  rootActor: AnyActorRef,
+  options: {
+    clock: Clock;
+    snapshot?: unknown;
```
Comment (Andarist) [inline packages/core/src/system.ts:92]: this stays unused but we will need this in the future so I'm OK with keeping it here

File: packages/core/src/system.ts [hunk 10]
```diff
@@ -1,22 +1,109 @@
...
+  /**
+   * @internal
+   */
+  _relay: (
+    source: AnyActorRef | undefined,
+    target: AnyActorRef,
+    event: AnyEventObject
+  ) => void;
+  scheduler: Scheduler;
+  getSnapshot: () => {
```
[the remark this answers is not in the document]
Comment (davidkpiano) [reply packages/core/src/system.ts:73]: We'll figure it out later; it's for our internal use only for now

File: packages/core/src/scheduler.ts (hunk 5 above)
Comment (Andarist) [reply packages/core/src/scheduler.ts]: done

File: packages/core/src/scheduler.ts (hunk 6 above)
Comment (Andarist) [reply packages/core/src/scheduler.ts]: we'll be doing the actual timers persistence later

File: packages/core/src/system.ts [hunk 11]
```diff
@@ -1,22 +1,182 @@
...
+  }
+
+  return systemPath;
+}
+
+function createScheduledEventId(
+  actorRef: AnyActorRef,
+  id: string
+): ScheduledEventId {
+  return `${getSystemPath(actorRef)}.${id}` as ScheduledEventId;
```
Comment (Andarist) [inline packages/core/src/system.ts:58]: Couldn't this be simplified like this?
```suggestion
  return `${actorRef.sessionId}.${id}` as ScheduledEventId;
```

**Review by Andarist — APPROVED** [review]

**Pull request merged**